dbms/src/Core/TaskOperatorSpillContexts.h

```diff
@@ -0,0 +1,100 @@
+
+namespace DB
+{
+class TaskOperatorSpillContexts
+{
+public:
+    Int64 triggerAutoSpill(Int64 expected_released_memories)
+    {
+        if (isFinished())
+            return expected_released_memories;
+        appendAdditionalOperatorSpillContexts();
+        bool has_finished_operator_spill_contexts = false;
+        for (auto & operator_spill_context : operator_spill_contexts)
+        {
+            assert(operator_spill_context->supportAutoTriggerSpill());
+            if (operator_spill_context->spillableStageFinished())
+            {
+                has_finished_operator_spill_contexts = true;
+                continue;
+            }
+            expected_released_memories = operator_spill_context->triggerSpill(expected_released_memories);
+            if (expected_released_memories <= 0)
+                break;
+        }
+        if (has_finished_operator_spill_contexts)
+        {
+            /// clean finished spill context
+            operator_spill_contexts.erase(std::remove_if(operator_spill_contexts.begin(), operator_spill_contexts.end(), [](const auto & context) { return context->spillableStageFinished(); }), operator_spill_contexts.end());
+        }
+        return expected_released_memories;
+    }
+    void registerOperatorSpillContext(const OperatorSpillContextPtr & operator_spill_context)
+    {
+        if (operator_spill_context->supportAutoTriggerSpill())
+        {
+            std::unique_lock lock(mutex);
+            operator_spill_contexts.push_back(operator_spill_context);
+            has_additional_operator_spill_contexts = true;
+        }
+    }
+    /// for tests
+    size_t operatorSpillContextCount()
+    {
+        appendAdditionalOperatorSpillContexts();
+        return operator_spill_contexts.size();
+    }
+
+    Int64 totalRevocableMemories()
+    {
+        if unlikely (isFinished())
+            return 0;
+        appendAdditionalOperatorSpillContexts();
+        Int64 ret = 0;
+        for (const auto & operator_spill_context : operator_spill_contexts)
+            ret += operator_spill_context->getTotalRevocableMemory();
+        return ret;
+    }
+
+    bool isFinished() const { return is_task_finished; }
+
+    void finish() { is_task_finished = true; }
+
+private:
+    void appendAdditionalOperatorSpillContexts()
+    {
+        if (has_additional_operator_spill_contexts)
+        {
+            std::unique_lock lock(mutex);
+            operator_spill_contexts.splice(operator_spill_contexts.end(), additional_operator_spill_contexts);
+            has_additional_operator_spill_contexts = false;
+            additional_operator_spill_contexts.clear();
+        }
+    }
+    /// access to operator_spill_contexts is thread safe
+    std::list<OperatorSpillContextPtr> operator_spill_contexts;
+    mutable std::mutex mutex;
+    /// access to additional_operator_spill_contexts need acquire lock first
+    std::list<OperatorSpillContextPtr> additional_operator_spill_contexts;
```

Comment:
where is additional_operator_spill_contexts written?

Reply:
In pipeline model, the spill context will be created at runtime, then it will be added on created, in pull model, if join spill is triggered, the hash join spill context will also be created at runtime